### PR TITLE
[doc] Add architectural concerns to the review checklist

### PR DESCRIPTION
**Pull request by sbegaudeau**
Signed-off-by: Stéphane Bégaudeau <stephane.begaudeau@obeo.fr>

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/review-checklist.adoc [hunk 1]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
 
+## Code
+
+### Architecture
+
+This repository contains two main parts: Sirius Components and Sirius Web.
+
+#### Sirius Components
+
+The modules of Sirius Components are to be reused by countless of applications in various use cases.
```
Comment (frouene) [inline doc/review-checklist.adoc]: are meant to be reused ?

File: doc/review-checklist.adoc [hunk 2]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
+
+Here are some use cases of applications in which a subset of Sirius Components should be usable:
+
+- *Java applications not using the Spring runtime*:
+We should be able to embed our representations in a Java application based on another framework than Spring.
+Parts of our code can have dependencies to Spring annotations like `@Service` but we need to isolate our dependency to Spring.
+Yet, we should not rely more than necessary on Spring to build our API and behavior.
+- *Applications not relying on an internet connection:*
+We have demonstrated in the past that both the frontend and backend modules of Sirius Components can be embedded in a VS Code extension which works locally.
+It still uses an internal HTTP / WebSocket server to communicate between the frontend and the backend by does not rely on an internet connection to a remote server.
```
Comment (frouene) [inline doc/review-checklist.adoc]: typo on "by does" ?

File: doc/review-checklist.adoc [hunk 3]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
 
 Issues are quite simple to review, you should mostly consider if the appropriate labels have been added with at least a label for:
 
-- The area such as `area: backend`, `area: frontend`, `area: diagram` etc
+- The various packages that will be impacted such as `package: diagram`, `package: form`, `package: view` etc
 - The difficulty such as `difficulty: starter`
-- The type like `type: enhancement``
+- The type like `type: enhancement`
 
 Those various labels are important to find existing issues and also because we have from time to time some contributors available on the project for a couple of day or weeks and we need to find simple enhancements for them to perform.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: of days

File: doc/review-checklist.adoc [hunk 4]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
...
 
 Do not add try to create labels of various shapes in the title of your issues such as `Topic:` or `[Topic]`.
-If it's necessary, ask for the creation of a proper label dedicated to your topic.
+If it's necessary, ask for the creation of a proper label dedicated to your topic if the topic represents a sizable part of the project long term.
 It will make it easier to find issues in the future.
 
-Issues and pull requests on Sirius Components should be focused on Sirius Components.
-If an application integrating Sirius Components needs to be mentionned, please use Sirius Web.
+Issues and pull requests in this repository should be focused on Sirius Components and Sirius Web.
+If an application integrating Sirius Components needs to be mentionned, please use Sirius Web as an example as much as possible.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: mentioned

File: doc/review-checklist.adoc [hunk 5]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
```
Comment (mcharfadi) [inline doc/review-checklist.adoc:8]: contain a cheklist --> checklist

File: doc/review-checklist.adoc [hunk 6]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
...
-If you know for a fact that your work will be part of the current iteration, you can add the current project and milestone to the issue and pull request.
+If you know for a fact that some work will be part of the current iteration, you can add the current project and milestone to the issue and pull request.
+
+IMPORTANT: Since we are managing our work using the Github project, all issues associated with the current milestone should be visible in the project too.
 
 
 ### Linked pull requests
 
 If there are pull requests for an issue, you should link them.
 Once the pull request will be merged, the issue will be closed automatically.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: will be merged --> is merged

File: doc/review-checklist.adoc [hunk 7]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
...
+It is also essential in order to have a bird eye view of the work being done in the project.
 If some work remains to be done, it's best to open a new issue to indicate that a first part of the work has been merged and that some work remains to be done.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: a first --> the first or just part of

File: doc/review-checklist.adoc [hunk 8]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
...
+Large scale / generic issues tends to be problematic since you never know if you can close them since the work may not be fully completed.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: tends --> tend

File: doc/review-checklist.adoc [hunk 9]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
...
+Since issues can only be affected to a single milestone, it creates issues when some work is spanning accross multiple iterations.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: affected by &
across

File: doc/review-checklist.adoc [hunk 10]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
...
 
 ## Pull requests
 
 ### Commit
 
-All the commits of a pull request should be properly formatted.
+We are expecting to frequently merge meaningful improvements to both Sirius Components and Sirius Web.
+As a result, the most important metric to keep delivering meaningful improvements is the size of the scope of each improvement.
+You can organize your work as you want with any number of commits to remember what you have done but when it is time to submit your work to be reviewed, it is best to have the lowest number of commits possible.
+Since a contribution is expected to small in scope, most of the pull requests of the project contain a single commit.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: to be small

File: doc/review-checklist.adoc [hunk 11]
```diff
@@ -8,41 +8,62 @@ This document will contain a cheklist of the various things to consider while su
...
+We frequently have pull requests with two or three commits with, for example, a commit for an ADR or a commit for a small isolated improvement not strictly related to the main work but most of the time, the smaller a pull request is, the better it is for everybody.
+
+All the commit messages of a pull request should be properly formatted.
 Have a look at other commits if you have any doubts about the way we format our commit messages.
-Commits should thus start with a title starting with the topic of the commit, either the number of an issue or a general purpose tag like `[releng]`, `[test]`, `[doc]` etc.
+Commits should thus start with a title starting with the number of the issue linked to the pull request most of the time or a general purpose tag like `[releng]`, `[test]`, `[doc]` etc.
 General purpose tags should only be used for small improvements.
 
+TIP: While an issue should describe the problem to be fixed or the feature to implement, commit messages should describe the solution implemented.
+Countless of issues would benefit from having a commit message with a tiny bit more of information than just copying and pasting the title of the issue.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: Countless issues

File: doc/review-checklist.adoc [hunk 12]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
 The code should have a proper copyright which should be updated when necessary.
 
 A commit fixing an issue should also update the `CHANGELOG` to indicate what has been done.
-For most contributions, a simple sentence with the link of the issue and it's description is enough.
+For most contributions, a simple sentence indicating what has been done with the link of the issue is enough.
 Important API breaks should also be noted in the dedicated sections.
 
+IMPORTANT: If a contribution changes code in Sirius Web, most of the time it is to adapt the code to a change in Sirius Components.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc:82]: a contributor

File: doc/review-checklist.adoc [hunk 13]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
+You can pretty much consider any contribution with a change in Sirius Web as having API breaks which should be described separately than the work being done in the CHANGELOG.
+
 Major evolutions require a dedicated `ADR` which should be referenced in the `CHANGELOG` too.
 It is better to contribute the ADR before starting to work on an issue in order to ensure that you are not starting in a wrong direction.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc:86]: contribute to the ADR

File: doc/review-checklist.adoc [hunk 14]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
 
+## Code
+
+### Architecture
+
+This repository contains two main parts: Sirius Components and Sirius Web.
+
+#### Sirius Components
+
```
Comment (mcharfadi) [inline doc/review-checklist.adoc:95]: countless applications

File: doc/review-checklist.adoc [hunk 15]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
+It still uses an internal HTTP / WebSocket server to communicate between the frontend and the backend by does not rely on an internet connection to a remote server.
+- *Java applications not using any frontend at all:*
+We should be able to create representations fully in Java without any JavaScript/CSS/HTML at all.
+The application integrating Sirius Components module would have to figure out how to render the representation for its users but it should be possible.
+- *Java applications without Eclipse EMF:*
+In some use cases, our consumer may not leverage EMF to manipulate their data.
+As such, the core parts of our application should not consider that EMF will be available.
+- *Java applications without odesign or our View DSL:*
+They would have to manipulate bare metal APIs without the refinements and high level concepts of our View DSL but it has advantages too.
+Fully programmatic manipulation can be easier to maintain for experienced Java developers and more performant for advanced team.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: teams

File: doc/review-checklist.adoc [hunk 16]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
+They should not have any opinion on the persistence or network solutions adopted by the application in which they are integrated.
+Those modules should not rely on advanced Spring features since they will be reused in non-Spring applications.
+They can depend on Spring annotations like `@Service` but not much more.
+- Maven modules used to provide support for the GraphQL API should be independent from Eclipse EMF, the View DSL, the persistence solution or even the network protocols used (HTTP or WebSocket or Server Sent Event etc.)
+- Our NPM packages should not make any hypothesis on the way the representations have been created (odesign, View DSL, programmatically).
+They should not either be dependent on a specific protocol apart from HTTP to fetch images to be displayed in a regular image element like `<img src="..." />`.
+Some application could decide that our GraphQL queries and mutations could be sent over WebSocket and our subscriptions sent using Server Sent Events.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: applications

File: doc/review-checklist.adoc [hunk 17]
```diff
@@ -55,12 +76,85 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
+It should never be considered as the sole use case or even the main use case in which the end users are interacting with Sirius Components.
+
+The goal of Sirius Web is to provide a simple web application integrating Sirius Components.
+Since a lot of consumers of Sirius Components will want to create web applications like Sirius Web, it has to provide some answers on how Sirius Components can behave in a web environment.
+For that, Sirius Web will have to solve specific issues outside of the scope of Sirius Components.
+Those solutions will thus force Sirius Web to depend on additional technologies and introduce additional concepts.
+
+IMPORTANT: *Sirius Web specific concepts and dependencies should not leak into Sirius Components code*.
+
+Here are a couple of example of concerns specific to Sirius Web.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc]: examples

File: doc/review-checklist.adoc [hunk 18]
```diff
@@ -75,7 +169,7 @@ Here is a list of things to consider while reviewing some backend changes:
 - [ ] Are there non `final` fields and are they really necessary?
```
Comment (mcharfadi) [inline doc/review-checklist.adoc:169]: Typo at Are there new depencencies?

File: doc/review-checklist.adoc [hunk 19]
```diff
@@ -75,7 +169,7 @@ Here is a list of things to consider while reviewing some backend changes:
...
 - [ ] Does a data structure contain some behavior?
 - [ ] Is any data structure mutable? Is that really necessary?
-- [ ] Is the dependency to EMF / Sirius RCP properly isolated?
+- [ ] Are the dependencies relevant and up to date?
 
 
 ### Frontend
```
Comment (mcharfadi) [inline doc/review-checklist.adoc:175]: This is continuous --> This continuous

Comment (mcharfadi) [reply doc/review-checklist.adoc:175]: Same typo below at "Are there new depencencies?"

File: doc/review-checklist.adoc [hunk 20]
```diff
@@ -107,7 +201,7 @@ The reviewer should consider if the PR needs to be squashed and he should do it
 
 After that, the reviewer should use the `Rebase and merge` action to merge the PR and `Delete the branch` to clean the, now useless, branch.
 
-If the commit is necessary for another project, for example to merge some changes in Sirius Web, then a new release should be performed.
+If the commit is necessary for another project, then a new release should be performed.
```
Comment (mcharfadi) [inline doc/review-checklist.adoc:204]: below : this preventS the need

File: doc/review-checklist.adoc (hunk 12 above)
Comment (sbegaudeau) [reply doc/review-checklist.adoc:82]: I really meant contribution not contributor

File: doc/review-checklist.adoc (hunk 13 above)
Comment (sbegaudeau) [reply doc/review-checklist.adoc:86]: No I meant contributing the ADR relevant to the issue.

File: doc/review-checklist.adoc [hunk 21]
```diff
@@ -55,40 +76,113 @@ It's not required anymore by the Eclipse Foundation and we may remove this requi
...
+
+Here are some use cases of applications in which a subset of Sirius Components should be usable:
+
+- *Java applications not using the Spring runtime*:
+We should be able to embed our representations in a Java application based on another framework than Spring.
+Parts of our code can have dependencies to Spring annotations like `@Service` but we need to isolate our dependency to Spring.
+Yet, we should not rely more than necessary on Spring to build our API and behavior.
+- *Applications not relying on an internet connection:*
+We have demonstrated in the past that both the frontend and backend modules of Sirius Components can be embedded in a VS Code extension which works locally.
+It still uses an internal HTTP / WebSocket server to communicate between the frontend and the backend does not rely on an internet connection to a remote server.
```
Comment (AxelRICHARD) [inline doc/review-checklist.adoc]: `... between the frontend and the backend AND does not rely on...`?
or may be `... between the frontend and the backend AND THE BACKEND does not rely on...`?

Comment (sbegaudeau) [reply doc/review-checklist.adoc]: I've added a "but" :)

**Review by gcoutable — APPROVED** [review]

**Pull request merged**